common/txmgr/finalizer.go

```diff
@@ -0,0 +1,235 @@
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("failed to find latest finalized head in chain")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	confirmedTxs, err := f.txStore.FindTransactionsByState(ctx, TxConfirmed, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve confirmed transactions: %w", err)
+	}
+
+	var finalizedTxs []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	// Group by block hash transactions whose receipts cannot be validated using the cached heads
+	receiptBlockHashToTx := make(map[BLOCK_HASH][]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE])
+	// Find transactions with receipt block nums older than the latest finalized block num and block hashes still in chain
+	for _, tx := range confirmedTxs {
+		// Only consider transactions not already marked as finalized
+		if tx.Finalized {
+			continue
+		}
+		receipt := tx.GetReceipt()
+		if receipt == nil || receipt.IsZero() || receipt.IsUnmined() {
+			continue
+		}
+		// Receipt newer than latest finalized head block num
+		if receipt.GetBlockNumber().Cmp(big.NewInt(latestFinalizedHead.BlockNumber())) > 0 {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.GetBlockNumber().Int64() < earliestBlockNumInChain {
+			receiptBlockHashToTx[receipt.GetBlockHash()] = append(receiptBlockHashToTx[receipt.GetBlockHash()], tx)
+			continue
+		}
+		blockHashInChain := latestFinalizedHead.HashAtHeight(receipt.GetBlockNumber().Int64())
+		// Receipt block hash does not match the block hash in chain. Transaction has been re-org'd out but DB state has not been updated yet
+		if blockHashInChain.String() != receipt.GetBlockHash().String() {
+			continue
+		}
+		finalizedTxs = append(finalizedTxs, tx)
+	}
+
+	// Check if block hashes exist for receipts on-chain older than the earliest cached head
+	// Transactions are grouped by their receipt block hash to minimize the number of RPC calls in case transactions were confirmed in the same block
+	// This check is only expected to be used in rare cases if there was an issue with the HeadTracker or if the node was down for significant time
+	var wg sync.WaitGroup
+	var txMu sync.RWMutex
+	for receiptBlockHash, txs := range receiptBlockHashToTx {
+		wg.Add(1)
+		go func(hash BLOCK_HASH, txs []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+			defer wg.Done()
+			if head, rpcErr := f.client.HeadByHash(ctx, hash); rpcErr == nil && head.IsValid() {
```

Comment:
I would also check if the block number of the requested head is lower than the fetched `latestFinalizedHead` just to be on the safe side, otherwise we could throw an invariant violation error.

Reply:
I'm might be talking about a different thing, but I agree that we need to check fetched block against the latest finalized block, but we should use HeadTracker's `LatestAndFinalizedBlock` from [this PR](https://github.com/smartcontractkit/chainlink/pull/13048).

The main reason is that MultiNode only provides a repeatable read guarantee for data that is older than the finalized block fetched with `LatestAndFinalizedBlock`. It does not provide this guarantee for blocks fetched by `LatestFinalizedHead`.  
Why? Because of HeadTracker's FinalityTagBypass option. We plan to remove it with HeadTracker's performance optimization.

Reply:
I added a check to ensure the fetched block's num is still earlier or equal to the latest finalized block num. Thanks for the heads up! I'll update to using `LatestAndFinalizedBlock` once your PR is merged. For now, I've marked this PR dependent on yours.